### PR TITLE
[Autoscaler] Multi node commands

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR adds support for per node `initialization_commands`, `setup_commands`, etc

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by rkooo567** [comment]
@wuisawesome Lint failure.

**Comment by wuisawesome** [comment]
whoops fixed

File: python/ray/autoscaler/autoscaler.py [hunk 1]
```diff
@@ -389,6 +390,19 @@ def recover_if_needed(self, node_id, now):
         updater.start()
         self.updaters[node_id] = updater
 
+    def _get_commands(self, node_id: str, commands_key: str):
+        commands = self.config[commands_key]
+        node_tags = self.provider.node_tags(node_id)
+        if TAG_RAY_USER_NODE_TYPE in node_tags:
+            node_type = node_tags[TAG_RAY_USER_NODE_TYPE]
+            assert node_type in self.available_node_types,\
+                "Unknown node type tag: {}.".format(
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py:399]: Don't assert, raise ValueError instead. Asserts should be for internal logic asserts and not checking user input.

File: python/ray/autoscaler/autoscaler.py [hunk 2]
```diff
@@ -389,6 +390,19 @@ def recover_if_needed(self, node_id, now):
         updater.start()
         self.updaters[node_id] = updater
 
+    def _get_commands(self, node_id: str, commands_key: str):
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py:393]: Can you add pydoc for this method?

Comment (ericl) [inline python/ray/autoscaler/autoscaler.py:393]:
```suggestion
    def _get_commands_for_node_type(self, node_id: str, commands_key: str):
```

File: python/ray/autoscaler/commands.py [hunk 3]
```diff
@@ -455,6 +455,20 @@ def warn_about_bad_start_command(start_commands):
             "to ray start in the head_start_ray_commands section.")
 
 
+def _get_head_commands(config, commands_key):
```
Comment (ericl) [inline python/ray/autoscaler/commands.py:458]: Pydoc

File: python/ray/autoscaler/commands.py [hunk 4]
```diff
@@ -455,6 +455,20 @@ def warn_about_bad_start_command(start_commands):
...
+    commands = config[commands_key]
+    node_type = config.get("head_node_type")
+    if node_type:
+        assert "available_node_types" in config
+        assert node_type in config[
```
Comment (ericl) [inline python/ray/autoscaler/commands.py:463]: Prefer ValueError

File: python/ray/autoscaler/ray-schema.json [hunk 5]
```diff
@@ -291,6 +291,33 @@
...
+                       },
+                       "initialization_commands": {
+                           "$ref": "#/definitions/commands",
+                           "description": "List of commands that will be run before `setup_commands`. If docker is enabled, these commands will run outside the container and before docker is setup."
+                       },
+                       "setup_commands": {
+                           "$ref": "#/definitions/commands",
+                           "description": "List of common shell commands to run to setup nodes."
+                       },
+                       "head_setup_commands": {
```
Comment (ericl) [inline python/ray/autoscaler/ray-schema.json:303]: Do we really need head_setup / head_start_ray commands? I think we can just remove these and only allow them to be specified in the main config. After all there is only 1 head node ever.

File: python/ray/tests/test_resource_demand_scheduler.py [hunk 6]
```diff
@@ -306,6 +313,37 @@ def testResourcePassing(self):
...
+        autoscaler.update()
+        self.waitForNodes(1)
+        assert self.provider.mock_nodes[0].node_type == "m4.large"
+        autoscaler.request_resources([{"GPU": 8}])
+        autoscaler.update()
+        self.waitForNodes(2)
+        assert self.provider.mock_nodes[1].node_type == "p2.8xlarge"
+        autoscaler.update()
+        sleep(0.1)
+        runner.assert_has_call(self.provider.mock_nodes[1].internal_ip,
```
Comment (ericl) [inline python/ray/tests/test_resource_demand_scheduler.py:344]: Can you make multiple node types and check each has different commands?

File: python/ray/tests/test_resource_demand_scheduler.py [hunk 7]
```diff
@@ -306,6 +313,37 @@ def testResourcePassing(self):
         runner.assert_has_call("172.0.0.1", "CPU: 32")
         runner.assert_has_call("172.0.0.1", "GPU: 8")
 
+    def testCommandPassing(self):
+        config = MULTI_WORKER_CLUSTER.copy()
+        config["available_node_types"]["p2.8xlarge"][
+            "worker_setup_commands"] = ["echo hello world"]
```
Comment (ericl) [inline python/ray/tests/test_resource_demand_scheduler.py:319]: It would be good to test all the commands not just worker_setup.

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/autoscaler/ray-schema.json (hunk 5 above)
Comment (wuisawesome) [reply python/ray/autoscaler/ray-schema.json:303]: personally, i think api consistency is worth it here

Comment (ericl) [reply python/ray/autoscaler/ray-schema.json:303]: Thinking about this more, I think the reason why to remove it is that users may be confused about where to put their configs. Can we actually do the following:
- Remove head_setup_* from the per-node commands
- Also remove worker_setup_*.

This would make it 100% consistent, since you can only specify head in the base commands, and there is neither head nor worker setup commands in the per-node commands. Btw, this is also how the original design had the config.

File: python/ray/autoscaler/ray-schema.json [hunk 8]
```diff
@@ -291,6 +291,33 @@
...
+                       },
+                       "worker_setup_commands": {
+                           "$ref": "#/definitions/commands",
+                           "description": "Commands that will be run on worker nodes after common setup."
+                       },
+                       "head_start_ray_commands": {
+                           "$ref": "#/definitions/commands",
+                           "description": "Command to start ray on the head node. You shouldn't need to modify this."
+                       },
+                       "worker_start_ray_commands": {
```
Comment (ericl) [inline python/ray/autoscaler/ray-schema.json:315]: Please remove this.

File: python/ray/autoscaler/ray-schema.json [hunk 9]
```diff
@@ -291,6 +291,33 @@
...
+                       },
+                       "head_setup_commands": {
+                           "$ref": "#/definitions/commands",
+                           "description": "Commands that will be run on the head node after common setup."
+                       },
+                       "worker_setup_commands": {
+                           "$ref": "#/definitions/commands",
+                           "description": "Commands that will be run on worker nodes after common setup."
+                       },
+                       "head_start_ray_commands": {
```
Comment (ericl) [inline python/ray/autoscaler/ray-schema.json:311]: Please remove this.

File: python/ray/autoscaler/ray-schema.json [hunk 10]
```diff
@@ -291,6 +291,33 @@
...
+                       },
+                       "setup_commands": {
+                           "$ref": "#/definitions/commands",
+                           "description": "List of common shell commands to run to setup nodes."
+                       },
+                       "head_setup_commands": {
+                           "$ref": "#/definitions/commands",
+                           "description": "Commands that will be run on the head node after common setup."
+                       },
+                       "worker_setup_commands": {
```
Comment (ericl) [inline python/ray/autoscaler/ray-schema.json:307]: Please remove this.

File: python/ray/autoscaler/ray-schema.json [hunk 11]
```diff
@@ -291,6 +291,33 @@
...
+                       },
+                       "head_start_ray_commands": {
+                           "$ref": "#/definitions/commands",
+                           "description": "Command to start ray on the head node. You shouldn't need to modify this."
+                       },
+                       "worker_start_ray_commands": {
+                          "$ref": "#/definitions/commands",
+                          "description": "Command to start ray on worker nodes. You shouldn't need to modify this."
+                       },
+                       "no_restart": {
```
Comment (ericl) [inline python/ray/autoscaler/ray-schema.json:319]: Please remove this -- I think this flag can be propagated automatically without it being part of the user schema.

File: python/ray/tests/test_resource_demand_scheduler.py (hunk 6 above)
Comment (ericl) [reply python/ray/tests/test_resource_demand_scheduler.py:344]: Also, would be great to check it is overriding the base commands (check the base command is not run).

File: python/ray/autoscaler/commands.py [hunk 12]
```diff
@@ -630,10 +664,10 @@ def get_or_create_head_node(config,
                 init_commands = []
                 ray_start_commands = config["head_start_ray_commands"]
             elif no_restart:
-                init_commands = config["head_setup_commands"]
+                init_commands = _get_head_setup_commands(config)
```
Comment (ericl) [inline python/ray/autoscaler/commands.py:667]: As discussed before, can't we simply keep this as `head_setup_commands` (just ignore node specific setup)?

File: python/ray/autoscaler/commands.py [hunk 13]
```diff
@@ -630,10 +664,10 @@ def get_or_create_head_node(config,
...
                 ray_start_commands = []
             else:
-                init_commands = config["head_setup_commands"]
+                init_commands = _get_head_setup_commands(config)
```
Comment (ericl) [inline python/ray/autoscaler/commands.py:670]: Please revert this to just `head_setup_commands`.

File: python/ray/autoscaler/commands.py [hunk 14]
```diff
@@ -646,7 +680,8 @@ def get_or_create_head_node(config,
                 auth_config=config["auth"],
                 cluster_name=config["cluster_name"],
                 file_mounts=config["file_mounts"],
-                initialization_commands=config["initialization_commands"],
+                initialization_commands=_get_head_initialization_commands(
```
Comment (ericl) [inline python/ray/autoscaler/commands.py:683]: Please revert this to just `initialization_commands`.

File: python/ray/autoscaler/autoscaler.py [hunk 15]
```diff
@@ -389,6 +390,39 @@ def recover_if_needed(self, node_id, now):
...
+                return node_specific_config[key].copy()
+        return None
+
+    def _get_worker_initialization_commands(self, node_id):
+        from ray.autoscaler.docker import docker_pull_if_needed
+        default = self.config["initialization_commands"]
+        node_specific = self._get_node_specific_key(node_id,
+                                                    "initialization_commands")
+        if node_specific:
+            docker_pull_if_needed(self.config, node_specific)
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py:411]: I thought we were going to do these rewrites at parse time? We shouldn't need this here in that case.

File: python/ray/autoscaler/autoscaler.py [hunk 16]
```diff
@@ -389,6 +390,39 @@ def recover_if_needed(self, node_id, now):
...
+            return node_specific
+        else:
+            return default
+
+    def _get_worker_setup_commands(self, node_id):
+        from ray.autoscaler.docker import dockerize_worker_setup_if_needed
+        default = self.config["worker_setup_commands"]
+        node_specific = self._get_node_specific_key(node_id, "setup_commands")
+        if node_specific:
+            dockerize_worker_setup_if_needed(self.config, node_specific)
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py:421]: Similar question, can't this be done at parse time?

File: python/ray/autoscaler/commands.py (hunk 12 above)
Comment (wuisawesome) [reply python/ray/autoscaler/commands.py:667]: wait i thought we discussed node specific setup commands overriding both `head_setup_commands` and `setup_commands` (at the config level)

File: python/ray/autoscaler/docker.py [hunk 17]
```diff
@@ -8,21 +8,61 @@
...
+                                          docker_mounts, cname,
+                                          run_options + head_run_options)
 
-    worker_docker_image = config["docker"].get("worker_image", docker_image)
-    worker_run_options = config["docker"].get("worker_run_options", [])
+    return head_docker_start + (with_docker_exec(
+        config["head_setup_commands"], container_name=cname))
+
+
+def dockerize_if_needed(config):
```
Comment (ericl) [inline python/ray/autoscaler/docker.py:56]: This is a pretty big change to this file not related to this PR. Please try to keep diffs minimal (revert these changes). I think you only need to:
1) Add a for loop in this function that rewrites the type-specific commands as worker docker commands.
2) Add a second for loop for the type-specific init commands.

File: python/ray/autoscaler/commands.py (hunk 12 above)
Comment (ericl) [reply python/ray/autoscaler/commands.py:667]: Only for worker node, not for the head. The head node is never configured by per-node-type commands, which simplifies things a lot. This isn't a problem for the user since they can always put whatever head setup they want in the head setup command section.

File: python/ray/autoscaler/commands.py [hunk 18]
```diff
@@ -461,6 +461,40 @@ def warn_about_bad_start_command(start_commands):
             "to ray start in the head_start_ray_commands section.")
 
 
+def _get_head_specific_key(config, key):
```
Comment (ericl) [inline python/ray/autoscaler/commands.py:464]: Can undo these changes in this file now.

**Review by ericl — APPROVED** [review]
LGTM

**Comment by ericl** [comment]
Some lint failures: https://travis-ci.com/github/ray-project/ray/jobs/377715741
